### PR TITLE
add health check for load balance servers

### PR DESCRIPTION
**Pull request by samanhappy**
This PR introduces a health check mechanism for proxy servers which support configuring probe interval, path, timeout, fail and pass threhold.

### CODE REVIEW

File: pkg/filters/proxy/loadbalance.go [hunk 1]
```diff
@@ -126,30 +147,103 @@ func (h hasher) Sum64(data []byte) uint64 {
...
+	}
+	if spec.Fails == 0 {
+		spec.Fails = HealthCheckDefaultFailThrehold
+	}
+	if spec.Passes == 0 {
+		spec.Passes = HealthCheckDefaultPassThrehold
+	}
+	blb.probeClient = &http.Client{Timeout: time.Duration(spec.Timeout) * time.Second}
+	t := time.NewTicker(time.Duration(spec.Interval) * time.Second)
+	go func(t *time.Ticker) {
```
Comment (localvar) [inline pkg/filters/proxy/loadbalance.go:202]: this goroutine is leaked

File: pkg/filters/proxy/server.go [hunk 2]
```diff
@@ -31,6 +34,14 @@ type Server struct {
 	Weight         int      `json:"weight" jsonschema:"omitempty,minimum=0,maximum=100"`
 	KeepHost       bool     `json:"keepHost" jsonschema:"omitempty,default=false"`
 	addrIsHostName bool
+	health         atomic.Value
```
Comment (localvar) [inline pkg/filters/proxy/server.go:37]: is it necessary to use `atomic.Value`? it seems there are only one goroutine that accesses this field.

**Review by localvar — CHANGES_REQUESTED** [review]
when service discovery is enabled, the server list changes from time to time, this case is not handled by this pr.

File: doc/reference/filters.md [hunk 3]
```diff
@@ -1132,6 +1133,16 @@ Rules to revise request header.
 | lbCookieName | string | Name of the cookie generated by load balancer, its value will be used as the session identifier for stickiness in `DurationBased` and `ApplicationBased` mode, default is `EG_SESSION`             | No      |
 | lbCookieExpire | string | Expire duration of the cookie generated by load balancer, its value will be used as the session expire time for stickiness in `DurationBased` and `ApplicationBased` mode, default is 2 hours             | No      |
 
+### proxy.HealthCheckSpec
+
+| Name          | Type   | Description                                                                                                 | Required |
+| ------------- | ------ | ----------------------------------------------------------------------------------------------------------- | -------- |
+| interval | int | Interval seconds for health check | Yes |
+| path | string | Path URL for server health check | No |
+| timeout | int | Timeout seconds for health check, default is 3 | No |
```
Comment (localvar) [inline doc/reference/filters.md:1142]: please use duration format for `interval` and `timeout`

File: pkg/filters/proxy/loadbalance.go (hunk 1 above)
Comment (localvar) [reply pkg/filters/proxy/loadbalance.go:205]: the code change here haven't resolve the leak.

File: pkg/filters/proxy/basepool.go [hunk 4]
```diff
@@ -124,6 +131,10 @@ func (bsp *BaseServerPool) createLoadBalancer(spec *LoadBalanceSpec, servers []*
 		spec = &LoadBalanceSpec{}
 	}
 
+	if lb := bsp.LoadBalancer(); lb != nil {
+		lb.Close()
+	}
```
Comment (localvar) [inline pkg/filters/proxy/basepool.go:136]: the old instance should be closed after the new instance is ready.

File: pkg/filters/proxy/basepool.go [hunk 5]
```diff
@@ -125,7 +132,11 @@ func (bsp *BaseServerPool) createLoadBalancer(spec *LoadBalanceSpec, servers []*
 	}
 
 	lb := NewLoadBalancer(spec, servers)
+	old := bsp.LoadBalancer()
 	bsp.loadBalancer.Store(lb)
+	if old != nil {
+		old.Close()
+	}
```
Comment (localvar) [inline pkg/filters/proxy/basepool.go:139]: please use [SwapPointer](https://pkg.go.dev/sync/atomic#SwapPointer) to refactor L135-139.

File: pkg/filters/proxy/loadbalance.go [hunk 6]
```diff
@@ -259,6 +371,16 @@ func (blb *BaseLoadBalancer) ReturnServer(server *Server, req *httpprot.Request,
 	}
 }
 
+// Close closes resources
+func (blb *BaseLoadBalancer) Close() {
+	if blb.ticker != nil {
+		blb.ticker.Stop()
+	}
+	if blb.done != nil {
+		blb.done <- true
```
Comment (localvar) [inline pkg/filters/proxy/loadbalance.go:380]:
```suggestion
		close(blb.done)
```

File: pkg/filters/proxy/loadbalance.go [hunk 7]
```diff
@@ -126,30 +151,117 @@ func (h hasher) Sum64(data []byte) uint64 {
 type BaseLoadBalancer struct {
 	spec           *LoadBalanceSpec
 	Servers        []*Server
+	healthyServers atomic.Value
 	consistentHash *consistent.Consistent
 	cookieExpire   time.Duration
+	ticker         *time.Ticker
```
Comment (localvar) [inline pkg/filters/proxy/loadbalance.go:157]: no need to be a field of this struct, make it a local variable is better.
```suggestion
```

File: pkg/filters/proxy/loadbalance.go [hunk 8]
```diff
@@ -126,30 +151,117 @@ func (h hasher) Sum64(data []byte) uint64 {
...
+	go func() {
+		for {
+			select {
+			case <-blb.done:
+				return
+			case <-blb.ticker.C:
+				blb.probeServers()
+			}
+		}
+	}()
```
Comment (localvar) [inline pkg/filters/proxy/loadbalance.go:225]:
```suggestion
	ticker := time.NewTicker(blb.probeInterval)
	blb.done = make(chan bool)
	go func() {
		for {
			select {
			case <-blb.done:
			        ticker.Stop()
				return
			case <-ticker.C:
				blb.probeServers()
			}
		}
	}()
```

**Review by localvar — APPROVED** [review]

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**